testhelpers/docker_registry.go

```diff
@@ -88,6 +96,27 @@ func ReadOnly(handler http.Handler) http.Handler {
 	})
 }
 
+func Custom(handler http.Handler, permissions map[string]*ImagePrivileges) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if permission, ok := permissions[extractImageName(r.URL.Path)]; ok {
+			if r.Method == "GET" || r.Method == "HEAD" {
+				if !permission.readable {
+					w.WriteHeader(401)
+					_, _ = w.Write([]byte("Unauthorized.\n"))
```

Comment:
Looking at the return value of `ReadOnly` (although the use case is a bit different), I'm wondering whether it should be 401 or 405 ("Method Not Allowed")? I'm not sure which one is more accurate in this case.

Reply:
Well, that's actually one of the problems with the ReadOnly when we try to validate read/write access, we are expecting to receive `401 Unauthorized` if your credentials are wrong or you don't have access to resources.

Reply:
So why don't we fix the `ReadOnly` function?

Reply:
There are several tests that depend on that, it could be an option, but actually not sure how good or bad it could be, but in any case, this docker registry implementation is a custom implementation created just to handle test scenarios. From my perspective having a version of a registry that behaves in a certain way for given scenarios is not that bad

Reply:
IIUC we expect Docker Hub to return a 401, therefore the test registry should return a 401.

Reply:
@natalieparellano, it makes sense that it'll return 401. My question is why `ReadOnly` returns 405 🤔

Reply:
I am not sure - IIRC @micahyoung had done some investigation on what the failure response looks like for different registries. The [distribution spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) is a bit ambiguous on what specific code should be returned. I would not be opposed to updating `ReadOnly`, but I don't think it needs to block this PR.

Reply:
Of course.
Maybe it's worth opening a separate issue for it - fix the errors that should be returned in each case (so at least we'll be consistent).